skipLists.go

```diff
@@ -0,0 +1,215 @@
+		update[i].next[i] = newNode
+	}
+}
+
+func (sl *skipList) remove(value int) {
+	// Check for empty list
+	isEmpty := true
+	for i := 0; i < sl.maxNoOfLevel; i++ {
+		if sl.head.next[i] != nil {
+			isEmpty = false
+			break
+		}
+	}
+	if isEmpty {
+		fmt.Println("Cannot remove from empty list")
+		return
+	}
+
+	// Check if list has only one element and it's the target
+	isOneElement := true
+	firstValue := sl.head.next[0].value
+	for i := 1; i < sl.maxNoOfLevel; i++ {
+		if sl.head.next[i] != nil && sl.head.next[i].next[i] != nil {
+			isOneElement = false
+			break
+		}
+	}
+	if isOneElement && firstValue == value {
+		for i := 0; i < sl.maxNoOfLevel; i++ {
+			sl.head.next[i] = nil
+		}
+		sl.maxNoOfLevel = 1
+		fmt.Println("Element", value, "deleted successfully.")
+		return
+	}
+
+	current := sl.head
+	update := make([]*Node, sl.maxNoOfLevel)
+
+	// Traverse the list to find the position of the node to be deleted.
+	for i := sl.maxNoOfLevel - 1; i >= 0; i-- {
+		for current.next[i] != nil && current.next[i].value < value {
+			current = current.next[i]
+		}
+		update[i] = current
+	}
+
+	current = current.next[0] // Clear the reference to the current node.
+
+	if current != nil && current.value == value {
+		for i := 0; i < len(current.next); i++ {
+			update[i].next[i] = current.next[i]
+		}
+
+		for sl.maxNoOfLevel > 0 && sl.head.next[sl.maxNoOfLevel-1] == nil {
+			sl.maxNoOfLevel--
+		}
+		fmt.Println("Element", value, "deleted successfully.")
+	} else {
+		fmt.Println("Element:", value, "does not exist!")
+	}
+}
+
+func (sl *skipList) search(value int) bool {
+	// Check for empty list
+	isEmpty := true
+	for i := 0; i < sl.maxNoOfLevel; i++ {
+		if sl.head.next[i] != nil {
+			isEmpty = false
+			break
+		}
+	}
+	if isEmpty {
+		fmt.Println("Element", value, "not found in empty list.")
+		return false
+	}
+
+	current := sl.head
+
+	for i := len(sl.head.next) - 1; i >= 0; i-- {
+		// Keep moving forward at the current level
+		//if the next node's value is less than the search value.
+		for current.next[i] != nil && current.next[i].value < value {
+			current = current.next[i]
+		}
+	}
+
+	current = current.next[0]
+
+	if current != nil && current.value == value {
+		fmt.Println("Element", value, "found.")
+		return true
+	} else {
+		fmt.Println("Element", value, "not found.")
+		return false
+	}
+}
+
+func (sl *skipList) display() {
+	// Check for empty list
+	isEmpty := true
+	for i := 0; i < sl.maxNoOfLevel; i++ {
+		if sl.head.next[i] != nil {
+			isEmpty = false
+			break
+		}
+	}
+	if isEmpty {
+		fmt.Println("Skip List: (empty)")
+		return
+	}
+	fmt.Println("Skip List:")
+
+	for i := len(sl.head.next) - 1; i >= 0; i-- {
+		current := sl.head.next[i]
+
+		fmt.Printf("Level %d: ", i)
+
+		for current != nil {
+			fmt.Printf("%d ", current.value)
+			current = current.next[i]
+		}
+		fmt.Println()
+	}
+}
+
+func main() {
+	maxLevels := 10
+	skipList := newSkipList(maxLevels)
+
+	// Test empty list
```

Comment:
Instead of hard-coding maxNoOfLevel, consider dynamically adjusting it based on the number of elements in the skip list. Use the relationship  maxNoOfLevel = log 2 (number of elements)

Reply:
> Instead of hard-coding maxNoOfLevel, consider dynamically adjusting it based on the number of elements in the skip list. Use the relationship maxNoOfLevel = log 2 (number of elements)

The `maxNoOfLevel` is now dynamically adjusted based on the relationship `maxNoOfLevel = log2(number of elements)`, ensuring efficient use of levels as the skip list grows or shrinks.

Reply:
Yep, good job